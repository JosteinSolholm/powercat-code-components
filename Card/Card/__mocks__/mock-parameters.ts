/* istanbul ignore file */

import { IInputs } from "../generated/ManifestTypes";
import { MockEnumProperty, MockStringProperty, MockWholeNumberProperty } from "./mock-context";
import { MockDataSet } from "./mock-datasets";

export function getMockParameters(): IInputs {
  return {
    AccessibleLabel: new MockStringProperty(),
    Title: new MockStringProperty(),
    Subtitle: new MockStringProperty(),
    Description: new MockStringProperty(),
    Items: new MockDataSet([]),
    Alignment: new MockEnumProperty(),
    HeaderImage: new MockStringProperty(),
    Image: new MockStringProperty(),
    ImagePlacement: new MockEnumProperty(),
    Size: new MockEnumProperty(),
    TabIndex: new MockWholeNumberProperty(),
    Tooltip: new MockStringProperty(),
  };
}
