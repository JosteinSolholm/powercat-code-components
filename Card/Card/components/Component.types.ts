import type { CardProps } from "@fluentui/react-components";
import { IToolbarItem } from "./Toolbar/Component.types";

export interface CustomCardProps extends CardProps {
  key: string;
  visible?: boolean;
  title: string;
  subTitle: string;
  headerImage: string;
  image: string;
  description: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  items: IToolbarItem[];
  width: number;
  height: number;
  disabled: boolean;
  onSelected: (item?: IToolbarItem, onSelectCalled?: boolean) => void;
  getPopoverRoot: () => HTMLElement;
  onResize?: (height: number) => void;
  ariaLabel?: string;
  placePreview?: string;
  selectedValue?: string;
  accessibleLabel: string;
}
