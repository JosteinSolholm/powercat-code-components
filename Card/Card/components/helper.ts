export interface FileObject {
  fileName: string;
  fileSize: number;
  mimeType: string;
  fileContent: string;
  fileUrl: string;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function getUrlfromImage(previewImage: FileObject): string {
  return previewImage.fileUrl ?? "";
}

export function sizeLength(size: string | undefined): number {
  const sizeMapping: { [key: string]: number } = {
    small: 16,
    medium: 24,
    large: 32,
  };

  return sizeMapping[size ?? "medium"] ?? 24;
}
