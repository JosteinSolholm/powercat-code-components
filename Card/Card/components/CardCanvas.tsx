import * as React from "react";
import {
  Card,
  CardHeader,
  CardPreview,
  makeStyles,
  Body1,
  CardFooter,
  tokens,
  Caption1,
  shorthands,
} from "@fluentui/react-components";
import { CustomCardProps } from "./Component.types";
import { ToolbarComponent } from "./Toolbar/ToolbarComponent";
import useResizeObserver from "@react-hook/resize-observer";
import { sizeLength } from "./helper";
import { StringConstants } from "../ManifestConstants";

const useStyles = makeStyles({
  horizontalCardImage: {
    borderRadius: "4px",
    width: "64px",
    height: "64px",
  },
  headerImage: {
    borderRadius: "4px",
    maxWidth: "44px",
    maxHeight: "44px",
  },
  header: {
    width: "100%",
    overflow: "hidden",
  },
  headerTitle: {
    whiteSpace: "nowrap",
    overflow: "hidden",
    textOverflow: "ellipsis",
    maxWidth: "100%",
    ...shorthands.margin(0),
  },
  headerDescription: {
    display: "-webkit-box",
    WebkitLineClamp: 2,
    WebkitBoxOrient: "vertical",
    overflow: "hidden",
    textOverflow: "ellipsis",
    maxWidth: "100%",
    color: tokens.colorNeutralForeground3,
    ...shorthands.margin(0),
  },
  text: { ...shorthands.margin(0) },
});

export const CardCanvas: React.FC<CustomCardProps> = React.memo((props: CustomCardProps) => {
  const {
    accessibleLabel,
    title,
    subTitle,
    getPopoverRoot,
    items,
    image,
    placePreview,
    onSelected,
    description,
    orientation,
    headerImage,
    disabled,
    size,
    onResize,
    width,
    ...rest
  } = props;
  const styles = useStyles();
  const target = React.useRef<HTMLDivElement>(null);

  useResizeObserver(target, (entry) => {
    if (onResize) {
      // Added size padding along with the actual card height
      onResize(entry.contentRect.height + sizeLength(size));
    }
  });

  return (
    <Card
      size={size}
      ref={target}
      orientation={orientation}
      aria-label={accessibleLabel}
      style={{ width: width, height: "fit-content" }}
      {...(items.length <= 1 && !disabled ? { onClick: () => onSelected() } : {})}
      {...rest}
    >
      {placePreview === StringConstants.AboveHeader && image.length > 0 && (
        <CardPreview>
          <img alt="Preview" src={image} />
        </CardPreview>
      )}
      <CardHeader
        className={styles.header}
        style={{ overflow: "hidden" }}
        {...(title && {
          header: (
            <Body1>
              <b className={styles.headerTitle}>{title}</b>
            </Body1>
          ),
        })}
        {...(headerImage
          ? {
              image: (
                <img
                  className={orientation === "horizontal" ? styles.horizontalCardImage : styles.headerImage}
                  alt="Header logo"
                  src={headerImage}
                  style={{ borderRadius: "4px" }}
                />
              ),
            }
          : {})}
        {...(subTitle
          ? {
              description: <Caption1 className={styles.headerDescription}>{subTitle}</Caption1>,
            }
          : {})}
      />
      {placePreview === StringConstants.BelowHeader && image.length > 0 && (
        <CardPreview>
          <img alt="Preview" src={image} />
        </CardPreview>
      )}
      {description && orientation === "vertical" && (
        <p className={styles.text}> {description.length > 700 ? `${description.slice(0, 700)}...` : description}</p>
      )}
      {items.length > 0 && (
        <CardFooter>
          <ToolbarComponent
            layout={"before"}
            disabled={disabled}
            items={items}
            width={width}
            onSelected={onSelected}
            getPopoverRoot={getPopoverRoot}
          />
        </CardFooter>
      )}
    </Card>
  );
});
CardCanvas.displayName = "CardCanvas";
