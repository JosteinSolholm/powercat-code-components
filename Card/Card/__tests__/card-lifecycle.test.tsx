import { Card } from "..";
import { IInputs } from "../generated/ManifestTypes";
import { MockContext, MockState } from "../__mocks__/mock-context";
import { MockDataSet, MockEntityRecord } from "../__mocks__/mock-datasets";
import { getMockParameters } from "../__mocks__/mock-parameters";
import { mount } from "enzyme";
import { ItemColumns } from "../ManifestConstants";
import { FileObject } from "../components/helper";
global.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
}));

/*Dummy file values to test with */
const headerImage: FileObject = {
  fileName: "sample-document.pdf",
  fileSize: 1048576,
  mimeType: "application/pdf",
  fileContent: "JVBERi0xLjQKJaqrrK0KNCAwIG9iago8P...",
  fileUrl: "https://example.com/files/sample-document.pdf",
};

const previewImage: FileObject = {
  fileName: "image-file.jpg",
  fileSize: 512000,
  mimeType: "image/jpeg",
  fileContent: "/9j/4AAQSkZJRgABAQEAAAAAAAD/4QAYRXhpZgA...",
  fileUrl: "https://example.com/files/image-file.jpg",
};

describe("Card", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers();
  });
  afterEach(() => {
    for (let i = 0; i < document.body.children.length; i++) {
      if (document.body.children[i].tagName === "DIV") {
        document.body.removeChild(document.body.children[i]);
        i--;
      }
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    if ((setTimeout as any).mock) {
      jest.runOnlyPendingTimers();
      jest.useRealTimers();
    }
  });

  it("renders", () => {
    const { component, context, notifyOutputChanged } = createComponent();
    component.init(context, notifyOutputChanged);
    const cardElement = component.updateView(context);
    expect(cardElement).toMatchSnapshot();
  });

  it("check accessible label", async () => {
    const { component, context, notifyOutputChanged } = createComponent();
    component.init(context, notifyOutputChanged);
    const accessibleLabel = "cardComponent";
    context.parameters.AccessibleLabel.raw = accessibleLabel;
    const cardElement = component.updateView(context);
    const cardComponent = mount(cardElement);

    const card = cardComponent.find('div[role="group"]');
    expect(card.prop("aria-label")).toBe(accessibleLabel);
  });
});

it("checks the whole cardComponent is disabled", () => {
  const { component, context, notifyOutputChanged } = createComponent();
  context.mode.isControlDisabled = true;
  component.init(context, notifyOutputChanged);

  const cardElement = component.updateView(context);
  const cardComponent = mount(cardElement);

  const buttons = cardComponent.find("button");
  buttons.forEach((button) => {
    expect(button.prop("disabled")).toBe(true);
  });
});

function createComponent() {
  const component = new Card();
  const notifyOutputChanged = jest.fn();
  const context = new MockContext<IInputs>(getMockParameters());
  context.parameters.Title.raw = "Title";
  context.parameters.Subtitle.raw = "Subtitle";
  context.parameters.HeaderImage.raw = headerImage;
  context.mode.isVisible = true;
  context.parameters.Alignment.raw = "Vertical";
  context.parameters.ImagePlacement.raw = "Above header";
  context.parameters.Image.raw = previewImage;
  context.parameters.Items = new MockDataSet([
    new MockEntityRecord("1", {
      [ItemColumns.Key]: "toolAdd",
      [ItemColumns.DisplayName]: "Add",
      [ItemColumns.IconName]: "Add",
      [ItemColumns.IconStyle]: "Filled",
      [ItemColumns.Appearance]: "Primary",
      [ItemColumns.Visible]: true,
    }),
    new MockEntityRecord("2", {
      [ItemColumns.Key]: "toolEdit",
      [ItemColumns.DisplayName]: "Edit",
      [ItemColumns.IconName]: "Edit",
      [ItemColumns.IconStyle]: "Regular",
      [ItemColumns.Appearance]: "Subtle",
      [ItemColumns.Visible]: true,
    }),
    new MockEntityRecord("3", {
      [ItemColumns.Key]: "toolDelete",
      [ItemColumns.DisplayName]: "Delete",
      [ItemColumns.IconName]: "Delete",
      [ItemColumns.Disabled]: true,
      [ItemColumns.IconStyle]: "Filled",
      [ItemColumns.Appearance]: "Subtle",
      [ItemColumns.Visible]: true,
    }),
    new MockEntityRecord("4", {
      [ItemColumns.Key]: "toolInfo",
      [ItemColumns.DisplayName]: "Info",
      [ItemColumns.IconName]: "Info",
      [ItemColumns.Disabled]: false,
      [ItemColumns.IconStyle]: "Filled",
      [ItemColumns.Appearance]: "Subtle",
      [ItemColumns.Visible]: false,
    }),
  ]);

  const state = new MockState();
  const container = document.createElement("div");
  document.body.appendChild(container);
  return { component, context, container, notifyOutputChanged, state };
}
