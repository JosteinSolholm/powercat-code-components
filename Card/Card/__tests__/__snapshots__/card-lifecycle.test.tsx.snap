// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders 1`] = `
<Memo(CardCanvas)
  accessibleLabel=""
  ariaLabel=""
  description={null}
  getPopoverRoot={[Function]}
  headerImage="https://example.com/files/sample-document.pdf"
  height={-1}
  image="https://example.com/files/image-file.jpg"
  items={
    Array [
      Object {
        "appearance": "primary",
        "data": MockEntityRecord {
          "getNamedReference": [MockFunction],
          "id": "1",
          "values": Object {
            "ItemAppearance": "Primary",
            "ItemDisplayName": "Add",
            "ItemIconName": "Add",
            "ItemIconStyle": "Filled",
            "ItemKey": "toolAdd",
            "ItemVisible": true,
          },
        },
        "disabled": false,
        "iconName": "Add",
        "iconStyle": "Filled",
        "id": "1",
        "key": "toolAdd",
        "name": "Add",
        "tooltip": undefined,
        "visible": true,
      },
      Object {
        "appearance": "subtle",
        "data": MockEntityRecord {
          "getNamedReference": [MockFunction],
          "id": "2",
          "values": Object {
            "ItemAppearance": "Subtle",
            "ItemDisplayName": "Edit",
            "ItemIconName": "Edit",
            "ItemIconStyle": "Regular",
            "ItemKey": "toolEdit",
            "ItemVisible": true,
          },
        },
        "disabled": false,
        "iconName": "Edit",
        "iconStyle": "Regular",
        "id": "2",
        "key": "toolEdit",
        "name": "Edit",
        "tooltip": undefined,
        "visible": true,
      },
    ]
  }
  onResize={[Function]}
  onSelected={[Function]}
  orientation="vertical"
  placePreview="Above header"
  subTitle="Subtitle"
  title="Title"
  visible={true}
  width={-1}
/>
`;
